### PR TITLE
Open in Editor: Fix paths for Perforce repos

### PR DESCRIPTION
**Pull request by vdavid**
Fixes https://github.com/sourcegraph/sourcegraph/issues/41713

Added an "IF github OR gitlab OR bitbucket cloud THEN remove code host and owner from the url ELSE only remove the code host" logic.
The reasoning is that
- we know about these three particular code hosts that they include the owner in the URL which we don't want to include, to maintain backward compatibility
- we know that Perforce users don't want to remove any other piece from the URL than the code host.
- we don't know about each of [the rest of the code hosts](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/blob/internal/extsvc/types.go?L111%3A12=) so we want to stay on the safe side and not remove any part of the URL that the user might want.

If there are user complaints about this behavior (e.g. Bitbucket Server users flag that the owner is included in their URLs and it's always redundant) then we can add more to this list, either based on the URL itself, or by filtering based on `repository.externalRepository.serviceType`.

Note: I kept the logic of filtering based on URL rather than based on `repository.externalRepository.serviceType` because the URL contains all the info we currently need and the interface to the logic is smaller this way, which seemed more elegant.

## Test plan

Only the background logic changed and I've added a bunch of unit tests, it should be enough.

## App preview:

- [Web](https://sg-web-dv-fix-open-in-editor-for-perforce.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-zvkevuayhk.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Review by philipp-spiess — APPROVED** [review]
Looks good to me! Thanks for adding the tests, they really help clear things up 🙂

File: client/web/src/open-in-editor/build-url.ts [hunk 1]
```diff
@@ -5,6 +5,13 @@ import type { UIRangeSpec } from '@sourcegraph/shared/src/util/url'
 import type { EditorReplacements, EditorSettings } from './editor-settings'
 import { Editor, getEditor, supportedEditors } from './editors'
 
+export function buildRepoBaseNameAndPath(repoName: string, filePath: string | undefined): string {
+    const codeHostsWithOwnerInUrl = ['github.com', 'gitlab.com', 'bitbucket.org']
+    const repoNameIncludesOwner = codeHostsWithOwnerInUrl.some(url => repoName.startsWith(url + '/'))
+    const bareRepoNamePieces = repoName.split('/').slice(repoNameIncludesOwner ? 2 : 1)
+    return path.join(...[...bareRepoNamePieces, ...(filePath ? [filePath] : [])])
```
[the remark this answers is not in the document]
Comment (vdavid) [reply client/web/src/open-in-editor/build-url.ts]: Oh, yeah, thanks, @philipp-spiess, it is equivalent. To make 100% sure, I've done the change locally and the tests passed.

File: client/web/src/open-in-editor/build-url.ts [hunk 2]
```diff
@@ -5,6 +5,13 @@ import type { UIRangeSpec } from '@sourcegraph/shared/src/util/url'
 import type { EditorReplacements, EditorSettings } from './editor-settings'
 import { Editor, getEditor, supportedEditors } from './editors'
 
+export function buildRepoBaseNameAndPath(repoName: string, filePath: string | undefined): string {
+    const codeHostsWithOwnerInUrl = ['github.com', 'gitlab.com', 'bitbucket.org']
+    const repoNameIncludesOwner = codeHostsWithOwnerInUrl.some(url => repoName.startsWith(url + '/'))
```
Comment (taras-yemets) [inline client/web/src/open-in-editor/build-url.ts:10]: @vdavid, it seems I'm a bit late to the party (the PR has already been merged), but I'm not sure if it covers GitHub Enterprise, self-hosted GitLab/Bitbucket. What do you think?

Comment (vdavid) [reply client/web/src/open-in-editor/build-url.ts:10]: @taras-yemets The reason is that I don't know about the URL structure of self-hosted GitLab and Bitbucket Server.
As I mentioned in the PR description, these and others can be added if we're positive that the URLs in our search results of files on these code hosts always include the owner name in the URL. Until then, I wanted to be on the safe side and not remove any info that might be helpful for the user.

Do you have any knowledge that makes you positive that our search results of files on self-hosted GitLab and Bitbucket Server code hosts contain the owner in their URL?

Comment (philipp-spiess) [reply client/web/src/open-in-editor/build-url.ts:10]: > Do you have any knowledge that makes you positive that our search results of files on self-hosted GitLab and Bitbucket Server code hosts contain the owner in their URL?

I am certain that this is added for self-hosted GitLab at least. Unfortunately I think for all these platforms where we "don't 100% know", we should keep the behavior the same as it was before.

**Pull request merged**